### DISCUSSION TITLE
How to model Kitty configuration as a grammar?

### DISCUSSION THREAD

**Discussion by clo4** [tree-sitter/tree-sitter#2148] [Q&A] [answered]
I'm making a Tree-sitter grammar for the Kitty terminal's configuration "language".

The problem is that each directive has its own grammar. Some treat everything as a number (startup fails if the value isn't correct), some directives treat everything as a string, some have positional arguments with specific syntax (see `symbol_map` and `mouse_map` below).

`.config/kitty/kitty.conf`:
```kitty-conf
include current-theme.conf
globinclude kitty.d/**/*.conf

scrollback_pager less --chop-long-lines --RAW-CONTROL-CHARS +INPUT_LINE_NUMBER

mouse_map shift+left click grabbed,ungrabbed mouse_handle_click selection link prompt

symbol_map U+E0A0-U+E0A3,U+E0C0-U+E0C7 PowerlineSymbols

disable_ligatures always

font_size 11.0
cursor_beam_thickness 1.5
cursor_blink_interval -1
```

That's an example of correct syntax.

The main issue I have: because each directive has its own grammar, it means for a directive like `font_size`, the argument `none` is **syntactically invalid**.

I'm not sure how to model this in the grammar. It seems from the documentation that it's better to make the grammar more general, but I'm not sure you could make this grammar more general without making it *wrong*.
Things I've considered/tried:
- **Make a general `directive` rule**
  - Means invalid directives would parse correctly
  - Means invalid values would parse
    - `visual_window_select_characters 123` will **always** parse `123` as a string no matter what, so matching that as a number is always going to be wrong
- **Have one rule per directive**
  - Means having to update the grammar **every time** a new directive is added to Kitty
  - Requires specifying 168 directives manually (which I'm already more than half way to having done)

Does anyone more experienced have an opinion on what I should do about this? Is there a middle ground solution I'm missing? Is there a grammar with a similar problem I can reference?

**Comment by kovidgoyal** [comment, marked as answer]
@clo4 I suggest you leverage the infrastructure in kitty for parsing these to autogenerate entries for most of the directives.  See the main function in gen-config.py. It shows you how to iterate over all directives. Look in kitty/conf/generate.py which contains code to convert the definitions into python. Many directives are either choices, or numbers for which generating a grammar is obviously straightforward. That will leave the directives that require custom parsing and the key and mouse maps. That should cut down on the amount of manual work you have to do.

You can write a script to do this for you and run it with

kitty +launch myscript.py

**Reply by clo4** [reply]
Thank you Kovid, this is very helpful and actionable! Really appreciate it, and thank you for all the work you put into this great terminal :)

**Reply by kovidgoyal** [reply]
You are welcome and feel free to ask if you need some pointers about navigating the config parsing infrastructure.
